llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1009,8 +1015,16 @@ static void simplifyRecipes(VPlan &Plan, LLVMContext &Ctx) {
       Plan.getEntry());
   VPTypeAnalysis TypeInfo(Plan.getCanonicalIV()->getScalarType(), Ctx);
   for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(RPOT)) {
-    for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
-      simplifyRecipe(R, TypeInfo);
+    // Populate a Worklist, as simplifyRecipe might return a new recipe that we
```

Comment:
Currently, it seems something like the following change
```
    for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
      VPRecipeBase *NewR = simplifyRecipe(R, TypeInfo);
      while (NewR)
        NewR = simplifyRecipe(*NewR, TypeInfo);
    }
```
would suffice, as simplifyRecipe() currently creates at most one new recipe, and the order in which recipes are simplified does not matter (right?). If/when simplifying a recipe R may create multiple new recipes, as long as only one NewR will replace all of R's uses, it may suffice to return NewR, and perhaps call recursivelySimplifyRecipes(NewR), similar to recursivelyDeleteDeadRecipes().

Reply:
Actually, if you see #93398, `simplifyRecipe` creates two new recipes in one instance: even if only one NewR replaces all of R's uses, doesn't `simplifyRecipe` match the root recipe, and not sub-recipes? The order in which recipes are simplified should not matter, I think.

Reply:
Yes, that creation of two new recipes follows the suggestion in https://github.com/llvm/llvm-project/pull/89386#discussion_r1579120585, thanks! Hence **currently**.

Regarding the order in which recipes are visited when trying to simplify each:

Instead of simplifying (X && Y) || (X && !Y) directly into X, we can first simplify into (X && (Y || Z)) where Z is !Y, introducing two new recipes. Then the new recipe (Y || Z) should be visited and simplified into 'true' followed by visiting the other new recipe (X && 'true') and simplifying it into X. How will the two new recipes be simplified if they are visited in the other order?

In addition to the order above of newly introduced recipes among themselves, once a recipe is simplified and RAUW'd, it may be potential to try to simplify its users - preferably **after** trying to simplify all its new recipes? E.g., ((X && Y) || (X && !Y)) || !X). I.e., should new recipes be appended into the worklist at its **front** rather than at its **back**.

Reply:
> Instead of simplifying (X && Y) || (X && !Y) directly into X, we can first simplify into (X && (Y || Z)) where Z is !Y, introducing two new recipes. Then the new recipe (Y || Z) should be visited and simplified into 'true' followed by visiting the other new recipe (X && 'true') and simplifying it into X. How will the two new recipes be simplified if they are visited in the other order?

Ah, yes. Interesting caveat here: sub-recipes should be simplified before the root recipes, and hence added to the worklist before the root recipes, right?

Regarding front versus back, agreed. Another interesting caveat.

Regarding moving forward, I will update #93998 in the morning with the two noted caveats; does this patch look okay though, as a preparatory step? (It was originally split from #93998 on Florian's request).

Reply:
+1 for gradually splitting this off.
Suggest this patch temporarily updates simplifyRecipe() to return a single VPRecipeBase* (or none), to be replaced by a small vector of recipes in the subsequent patch along with evident use for it (i.e., simplifying (X && Y) || (X && Z) into (X && (Y || Z)). BTW, this pattern could introduce a single new recipe - by modifying the RAUW'd OR to use Y and Z instead of (X && Y) and (X && Z) - but recipes modified during simplification should also be (re)visited, along with new ones.)

This patch is missing tests, demonstrating the simplification of new recipes introduced when simplifying others, i.e., its non-NFC-ness.

The subsequent patch, extending the code snippet above and following Florian's comment below, could perhaps use a worklist as follows, which could also be outlined into recursivelySimplifyRecipes(NewR):
```
    for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
      SmallVector<VPRecipeBase *> NewRecipes = simplifyRecipe(R, TypeInfo);
      while (!NewRecipes.empty()) {
        VPRecipeBase *NewR = NewRecipes.pop_back_val();
        append_range(NewRecipes, simplifyRecipe(*NewR, TypeInfo));
      }
    }
```

Reply:
> to be replaced by a small vector of recipes in the subsequent patch along with evident use for it (i.e., simplifying (X && Y) || (X && Z) into (X && (Y || Z)). BTW, this pattern could introduce a single new recipe - by modifying the RAUW'd OR to use Y and Z instead of (X && Y) and (X && Z) - but recipes modified during simplification should also be (re)visited, along with new ones.

Sorry, I'm having difficulty understanding this comment. How can the pattern introduce a single new recipe? If we return the recipe `X && (Y || Z)`, how will a pattern like `X && true` or `X && false` suffice? In other words, how can we ensure that `Y || Z` is re-queued for simplification? If we return the recipe `Y || Z`, that recipe might be simplfied, but then how does the root recipe involving X get re-queued for simplification?

> This patch is missing tests, demonstrating the simplification of new recipes introduced when simplifying others, i.e., its non-NFC-ness.

While I agree that this patch is technically not an NFC, we only return the `VPWidenCastRecipe` as a candidate for further simplification, and other patterns are not broken up. In practice, is it possible to cook up a test where the newly-created `VPWidenCastRecipe` actually gets simplified further?

Reply:
> append_range(NewRecipes, simplifyRecipe(*NewR, TypeInfo))

Also, didn't we decide to add it to the front of the worklist, and not the back?

Reply:
> How can the pattern introduce a single new recipe? If we return the recipe X && (Y || Z), how will a pattern like X && true or X && false suffice? In other words, how can we ensure that Y || Z is re-queued for simplification? If we return the recipe Y || Z, that recipe might be simplfied, but then how does the root recipe involving X get re-queued for simplification?

I checked, and we do indeed need to return both the root recipe and sub-recipe.